core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -211,7 +211,13 @@ public TransportClientNodesService removeTransportAddress(TransportAddress trans
     }
 
     public <Response> void execute(NodeListenerCallback<Response> callback, ActionListener<Response> listener) {
-        List<DiscoveryNode> nodes = this.nodes;
+        List<DiscoveryNode> nodes;
+        synchronized (mutex) {
```

Comment:
I think the invariants should be simply that `closed` is true, `nodes` is cleared, the transport service has disconnected, and sampling is cancelled. I think specifying the relative order in which these take place as an invariant isn't clean.